python/ray/dashboard/modules/virtual_cluster/tests/test_virtual_cluster.py

```diff
@@ -0,0 +1,405 @@
+            assert replica_sets == virtual_cluster_replica_sets
+            return True
+        except Exception as ex:
+            logger.info(ex)
+            return False
+
+    # Create a new virtual cluster with exclusive allocation mode.
+    virtual_cluster_id = "virtual_cluster_1"
+    allocation_mode = "exclusive"
+    replica_sets = {"4c8g": 1, "8c16g": 1}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with less nodes.
+    replica_sets = {"4c8g": 1}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with more nodes.
+    replica_sets = {"4c8g": 1, "8c16g": 1}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with zero node (make it empty).
+    replica_sets = {}
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # `virtual_cluster_1` has released all nodes, so we can now
+    # create a new virtual cluster with two nodes.
+    virtual_cluster_id = "virtual_cluster_2"
+    replica_sets = {"4c8g": 1, "8c16g": 1}
+    # Using mixed allocation mode.
+    allocation_mode = "mixed"
+    wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # TODO: uncomment this when `MixedCluster::IsIdleNodeInstance` is fixed.
+    # Update the virtual cluster with less nodes.
+    # replica_sets = {"4c8g": 1}
+    # wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with more nodes.
+    # replica_sets = {"4c8g": 1, "8c16g": 1}
+    # wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+    # Update the virtual cluster with zero node (make it empty).
+    # replica_sets = {}
+    # wait_for_condition(_create_or_update_virtual_cluster, timeout=10)
+
+
+@pytest.mark.parametrize(
+    "ray_start_cluster_head",
+    [
+        {
+            "include_dashboard": True,
+        }
+    ],
+    indirect=True,
+)
+@pytest.mark.parametrize("allocation_mode", ["exclusive", "mixed"])
+def test_create_and_update_virtual_cluster_with_exceptions(
+    disable_aiohttp_cache, ray_start_cluster_head, allocation_mode
+):
+    cluster: Cluster = ray_start_cluster_head
+    assert wait_until_server_available(cluster.webui_url) is True
+    webui_url = cluster.webui_url
+    webui_url = format_web_url(webui_url)
+
+    # Add two nodes to the primary cluster.
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "4c8g"})
+    cluster.add_node(env_vars={"RAY_NODE_TYPE_NAME": "8c16g"})
+
+    # Create a new virtual cluster with a non-exist node type.
+    result = create_or_update_virtual_cluster(
+        webui_url=webui_url,
+        virtual_cluster_id="virtual_cluster_1",
+        allocation_mode=allocation_mode,
+        replica_sets={"16c32g": 1},
+        revision=0,
+    )
+    assert result["result"] is False
+    assert str(result["msg"]).endswith("No enough node instances to assign.")
```

Comment:
We may have to tell users which node type and how many nodes we lack, so that users know how to scale up the primary cluster or adapt the virtual cluster requirements.